### PR TITLE
Fix Datadog tracer

### PR DESCRIPTION
**Pull request by sharmaansh21**
1. Why is this pull request needed and what does it do?
This fix tries to address the issue in #2862 bumps datadog tracer

2. Which issues (if any) are related?
#2862 

3. Which documentation changes (if any) need to be made?
n/a

### CODE REVIEW

**Comment by sharmaansh21** [comment]
With these changes
```
 Handling new trace with 2 spans: [service:"coredns" name:"servedns_dns_1053_gameserver.REDATED.com." resource:"servedns:dns://:1053 gameserver.REDATED.com." traceID:4480105348796920755 spanID:4480105348796920755 start:1559742655000301960 duration:161333 meta:<key:"coredns.io/name" value:"gameserver.REDATED.com." > meta:<key:"coredns.io/rcode" value:"NOERROR" > meta:<key:"coredns.io/type" value:"A" > meta:<key:"system.pid" value:"30939" > metrics:<key:"_sample_rate" value:1 > metrics:<key:"_sampling_priority_rate_v1" value:1 > metrics:<key:"_sampling_priority_v1" value:1 > metrics:<key:"_sublayers.duration.by_service.sublayer_service:coredns" value:161333 > metrics:<key:"_sublayers.span_count" value:2 > metrics:<key:"_top_level" value:1 >  service:"coredns" name:"cache" resource:"cache" traceID:4480105348796920755 spanID:8394077515441798185 parentID:4480105348796920755 start:1559742655000328718 duration:125130 metrics:<key:"_sampling_priority_v1" value:1 > ]

```

**Comment by yongtang** [comment]
@sharmaanshul2102 I think the test failures are related. Could you take look?

**Comment by sharmaansh21** [comment]
```
=== RUN   TestSecondaryZoneTransfer
==================
WARNING: DATA RACE
Write at 0x00c000286a80 by goroutine 236:
  github.com/miekg/dns.(*response).Close()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:734 +0xc7
  github.com/coredns/coredns/request.(*ScrubWriter).Close()
      <autogenerated>:1 +0x62
  github.com/coredns/coredns/plugin/file.Xfr.ServeDNS.func1()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/xfr.go:38 +0x74
Previous read at 0x00c000286a80 by goroutine 119:
  github.com/miekg/dns.(*Server).serveTCPConn()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:518 +0x3fc
Goroutine 236 (running) created at:
  github.com/coredns/coredns/plugin/file.Xfr.ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/xfr.go:36 +0x58c
  github.com/coredns/coredns/plugin/file.File.ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/file.go:79 +0x13c0
  github.com/coredns/coredns/plugin/file.(*File).ServeDNS()
      <autogenerated>:1 +0x10e
  github.com/coredns/coredns/core/dnsserver.(*Server).ServeDNS()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:242 +0xbae
  github.com/coredns/coredns/core/dnsserver.(*Server).Serve.func1()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:102 +0xdc
  github.com/miekg/dns.HandlerFunc.ServeDNS()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:37 +0x51
  github.com/miekg/dns.(*Server).serveDNS()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:602 +0x4fe
  github.com/miekg/dns.(*Server).serveTCPConn()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:517 +0x3ee
Goroutine 119 (finished) created at:
  github.com/miekg/dns.(*Server).serveTCP()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:434 +0x33a
  github.com/miekg/dns.(*Server).ActivateAndServe()
      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:337 +0x29c
  github.com/coredns/coredns/core/dnsserver.(*Server).Serve()
      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:106 +0x200
  github.com/mholt/caddy.startServers.func1.1()
      /home/travis/gopath/pkg/mod/github.com/mholt/caddy@v1.0.0/caddy.go:795 +0x86
==================
--- FAIL: TestSecondaryZoneTransfer (0.01s)
    testing.go:809: race detected during execution of test
```
@yongtang @miekg I don't think so, but any pointers will be helpful here.

**Comment by miekg** [comment]
ah, this was a long standing race that I never fixed. Seems the time has come.


>```
>=== RUN   TestSecondaryZoneTransfer
>==================
>WARNING: DATA RACE
>Write at 0x00c000286a80 by goroutine 236:
>  github.com/miekg/dns.(*response).Close()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:734 +0xc7
>  github.com/coredns/coredns/request.(*ScrubWriter).Close()
>      <autogenerated>:1 +0x62
>  github.com/coredns/coredns/plugin/file.Xfr.ServeDNS.func1()
>      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/xfr.go:38 +0x74
>Previous read at 0x00c000286a80 by goroutine 119:
>  github.com/miekg/dns.(*Server).serveTCPConn()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:518 +0x3fc
>Goroutine 236 (running) created at:
>  github.com/coredns/coredns/plugin/file.Xfr.ServeDNS()
>      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/xfr.go:36 +0x58c
>  github.com/coredns/coredns/plugin/file.File.ServeDNS()
>      /home/travis/gopath/src/github.com/coredns/coredns/plugin/file/file.go:79 +0x13c0
>  github.com/coredns/coredns/plugin/file.(*File).ServeDNS()
>      <autogenerated>:1 +0x10e
>  github.com/coredns/coredns/core/dnsserver.(*Server).ServeDNS()
>      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:242 +0xbae
>  github.com/coredns/coredns/core/dnsserver.(*Server).Serve.func1()
>      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:102 +0xdc
>  github.com/miekg/dns.HandlerFunc.ServeDNS()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:37 +0x51
>  github.com/miekg/dns.(*Server).serveDNS()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:602 +0x4fe
>  github.com/miekg/dns.(*Server).serveTCPConn()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:517 +0x3ee
>Goroutine 119 (finished) created at:
>  github.com/miekg/dns.(*Server).serveTCP()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:434 +0x33a
>  github.com/miekg/dns.(*Server).ActivateAndServe()
>      /home/travis/gopath/pkg/mod/github.com/miekg/dns@v1.1.12/server.go:337 +0x29c
>  github.com/coredns/coredns/core/dnsserver.(*Server).Serve()
>      /home/travis/gopath/src/github.com/coredns/coredns/core/dnsserver/server.go:106 +0x200
>  github.com/mholt/caddy.startServers.func1.1()
>      /home/travis/gopath/pkg/mod/github.com/mholt/caddy@v1.0.0/caddy.go:795 +0x86
>==================
>--- FAIL: TestSecondaryZoneTransfer (0.01s)
>    testing.go:809: race detected during execution of test
>```
>@yongtang @miekg I don't think so, but any pointers will be helpful here.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/coredns/coredns/pull/2868#issuecomment-499368051
/Miek

--
Miek Gieben

File: plugin/trace/trace.go [hunk 1]
```diff
@@ -53,7 +54,7 @@ func (t *trace) OnStartup() error {
 		case "zipkin":
 			err = t.setupZipkin()
 		case "datadog":
-			err = t.setupDatadog()
+			t.setupDatadog()
```
Comment (miekg) [inline plugin/trace/trace.go:57]: just inline this function, seems short enough with these changes

**Review by miekg — COMMENTED** [review]
lgtm, small comment

**Comment by yongtang** [comment]
The change overall looks good to me, minus the inline of `t.setupDatadog()` part.

I will go ahead and merge this one and create a follow up PR to address @miekg 's comment.

/lgtm

File: plugin/trace/trace.go [hunk 2]
```diff
@@ -74,20 +75,9 @@ func (t *trace) setupZipkin() error {
...
-	host := strings.Split(t.Endpoint, ":")
-	config.AgentHostname = host[0]
-
-	if len(host) == 2 {
-		config.AgentPort = host[1]
-	}
-
-	tracer, _, err := ddtrace.NewTracer(config)
+func (t *trace) setupDatadog() {
+	tracer := opentracer.New(tracer.WithAgentAddr(t.Endpoint), tracer.WithServiceName(t.serviceName), tracer.WithDebugMode(true))
```
Comment (uthark) [inline plugin/trace/trace.go:79]: How can I disable debug mode for production?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Fix Datadog t..." ]
>@uthark commented on this pull request.
>
>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In plugin/trace/trace.go:
>
>> @@ -74,20 +75,9 @@ func (t *trace) setupZipkin() error {
>        return err
> }
>
>-func (t *trace) setupDatadog() error {
>-       config := ddtrace.NewConfiguration()
>-       config.ServiceName = t.serviceName
>-
>-       host := strings.Split(t.Endpoint, ":")
>-       config.AgentHostname = host[0]
>-
>-       if len(host) == 2 {
>-               config.AgentPort = host[1]
>-       }
>-
>-       tracer, _, err := ddtrace.NewTracer(config)
>+func (t *trace) setupDatadog() {
>+       tracer := opentracer.New(tracer.WithAgentAddr(t.Endpoint), tracer.WithServiceName(t.serviceName), tracer.WithDebugMode(true))
>
>How can I disable debug mode for production?

this should always be false, no?

**Pull request merged**